src/gt4py/analysis/passes.py

```diff
@@ -1195,30 +1199,40 @@ def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
 
         def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
             assert isinstance(node, gt_ir.StencilImplementation)
-            self.demotables: Dict[str, Optional[str]] = {
-                temp_field: None for temp_field in node.temporary_fields
-            }
-            """Dictionary mapping temporaries to their most recently referenced stage."""
+            self.demotables = set(node.temporary_fields)
             self.visit(node)
-            return set(self.demotables.keys())
-
-        def visit_Stage(self, node: gt_ir.Stage, **kwargs: Any) -> None:
-            kwargs["stage_name"] = node.name
-            self.generic_visit(node, **kwargs)
 
-        def visit_Assign(self, node: gt_ir.Assign, **kwargs: Any) -> None:
-            self.visit(node.value, **kwargs)
-            if node.target.name in self.demotables:
-                self.demotables[node.target.name] = kwargs["stage_name"]
-
-        def visit_FieldRef(self, node: gt_ir.FieldRef, **kwargs: Any) -> None:
-            field_name = node.name
-            if field_name in self.demotables:
-                assert self.demotables[field_name], f"Temporary {field_name} has no stage."
-                if kwargs["stage_name"] != self.demotables[field_name] or any(
-                    val != 0 for val in node.offset.values()
-                ):
-                    self.demotables.pop(field_name)
+            return self.demotables
+
+        def visit_Stage(self, node: gt_ir.Stage, **kwargs) -> None:
+            self.read_fields: Set[str] = set()
+            self.written_fields: Set[str] = set()
+            self.generic_visit(node)
+
+            # Any fields left that have only been read should be discarded.
+            # These have been read before (never having been) assigned
+            read_not_written = self.read_fields - self.written_fields
+            for name in read_not_written:
+                self.demotables.discard(name)
+
+        def visit_Assign(self, node: gt_ir.Assign, **kwargs) -> None:
+            self.visit(node.value, **kwargs, is_value=True)
+            self.visit(node.target, **kwargs, is_target=True)
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef, **kwargs) -> None:
+            # 1. is never used with an offset
+            if any(val != 0 for val in node.offset.values()):
+                self.demotables.discard(node.name)
+
+            # 2. is never read before assigned to
+            if kwargs.get("is_value", False) and node.name not in self.written_fields:
+                self.read_fields.add(node.name)
```

Comment:
Can’t we just `self.demotables.discard(node.name)` here, just have `self.written_fields.add(node.name)` in the if below and get rid of `self.read_fields` and the discarding code above (around line 1214) entirely? Maybe I also missed something…